### DISCUSSION TITLE
How to define an MLM input that depends on a preprocessing operation

### DISCUSSION THREAD

**Discussion by fmigneault** [crim-ca/mlm-extension#7] [Q&A] [answered]
With a STAC Item using MLM extension, it is possible to define a model input as such: 

```json
{
  "type": "Feature",
  "properties": {
    "mlm:input": [
      {
        "name": "EO Data",
        "bands": ["B01", "B02", "B03", "B04", "B08", "B8A", "B09", "B11", "B12", "NDVI", "NDWI1", "NDWI2"],
        "input": {
          "shape": [-1, 12, 1024, 1024],
          "dim_order": ["batch", "channel", "height", "width"],
          "data_type": "uint16"
        },
        "pre_processing_function": null
      }
    ],
    "raster:bands": [
      { "name": "B01", "common_name": "coastal", "nodata": 0, "data_type": "uint16", "bits_per_sample": 15, "spatial_resolution": 60, "scale": 0.0001, "offset": 0, "unit": "m"},
      { "name": "B02", "common_name": "blue", "nodata": 0, "data_type": "uint16", "bits_per_sample": 15, "spatial_resolution": 10, "scale": 0.0001, "offset": 0, "unit": "m"},
      { "name": "B03", "common_name": "green", "nodata": 0, "data_type": "uint16", "bits_per_sample": 15, "spatial_resolution": 10, "scale": 0.0001, "offset": 0, "unit": "m"},
      { "name": "B04", "common_name": "red", "nodata": 0, "data_type": "uint16", "bits_per_sample": 15, "spatial_resolution": 10, "scale": 0.0001, "offset": 0, "unit": "m"},
      { "name": "B05", "common_name": "rededge1", "nodata": 0, "data_type": "uint16", "bits_per_sample": 15, "spatial_resolution": 20, "scale": 0.0001, "offset": 0, "unit": "m"},
      { "name": "B06", "common_name": "rededge2", "nodata": 0, "data_type": "uint16", "bits_per_sample": 15, "spatial_resolution": 20, "scale": 0.0001, "offset": 0, "unit": "m"},
      { "name": "B07", "common_name": "rededge3", "nodata": 0, "data_type": "uint16", "bits_per_sample": 15, "spatial_resolution": 20, "scale": 0.0001, "offset": 0, "unit": "m"},
      { "name": "B08", "common_name": "nir", "nodata": 0, "data_type": "uint16", "bits_per_sample": 15, "spatial_resolution": 10, "scale": 0.0001, "offset": 0, "unit": "m"},
      { "name": "B8A", "common_name": "nir08", "nodata": 0, "data_type": "uint16", "bits_per_sample": 15, "spatial_resolution": 20, "scale": 0.0001, "offset": 0, "unit": "m"},
      { "name": "B09", "common_name": "nir09", "nodata": 0, "data_type": "uint16", "bits_per_sample": 15, "spatial_resolution": 60, "scale": 0.0001, "offset": 0, "unit": "m"},
      { "name": "B10", "common_name": "cirrus", "nodata": 0, "data_type": "uint16", "bits_per_sample": 15, "spatial_resolution": 60, "scale": 0.0001, "offset": 0, "unit": "m"},
      { "name": "B11", "common_name": "swir16", "nodata": 0, "data_type": "uint16", "bits_per_sample": 15, "spatial_resolution": 20, "scale": 0.0001, "offset": 0, "unit": "m"},
      { "name": "B12", "common_name": "swir22", "nodata": 0, "data_type": "uint16", "bits_per_sample": 15, "spatial_resolution": 20, "scale": 0.0001, "offset": 0, "unit": "m"}
    ]
  }
}
```

Provided that the Item also includes a `raster:bands` definition that describe the B01 to B12 bands, this is valid against the schema to define `mlm:input[0].bands` cross-references. 

However, in this specific case, the model also takes as input the NDVI, NDWI1 and NDWI2, which are themselves derived from a calculation combining the other bands: 
- `NDVI = (B08 - B04) / (B08 + B04)`
- `NDWI1 = (B03 - B08) / (B03 + B08)`
- `NDWI2 = (B08 - B11) / (B08 + B11)`

What would be the best way to represent those calculations to define those "virtual bands".

Things considered:
- https://github.com/stac-extensions/processing#expression-object with `processing:expression` and  `format: rio-calc` using above formulas
- https://github.com/stac-extensions/virtual-assets#fields with `vrt:algorithm` and `vrt:algorithm_opts` similar to what processing does

In both cases, the question remains, where to put the expression/algorithm? 
Given that this STAC Item is not itself a "data provider" definition listing Asset URIs, but rather a model "using/expecting" these Assets, it seems that nesting `processing:expression` or `vrt:algorithm`/`vrt:hrefs` in the `"assets"` section is not appropriate, as the source Assets are not offered by this STAC Item. That would be misleading for a client reading the contents of the Item. 

Another possibility is to define new "virtual bands" under `raster:bands`, which would contain a nested `processing:expression`. 
However, that seems to be more of a hack than anything else, since that would not be validated against the `processing` extension. Including the extension in `stac_extensions` would also cause an error (due to no `processing:...` field found), which would require a random field just to pass validation...

The `pre_processing_function` (https://github.com/crim-ca/mlm-extension#processing-expression) seems like the best candidate to receive this information and avoid stretching the definition of other extensions, but it is currently available only for working over the full MLM input (ie: the set of 9 bands + 3 virtual ones all at once). Since only some "bands" receive a preprocessing, this definition becomes somewhat convoluted under a single field.

Should an array be permitted in `pre_processing_function`, with indices of the array corresponding to per-band operations? 
Any other suggestions?

**Comment by fmigneault** [comment]
@m-mohr @emmanuelmathot @rbavery
I would like your input regarding this, to see if there are any other suggestions or considerations when involving the full STAC ecosystem.

**Reply by rbavery** [reply]
I think defining a more flexible processing field is the way to go, ideally we would re-use the `processing:expression` extension. Could we instead convert the `pre_processing_function` field into an array of `processing:expression` objects? This field could be called `mlm:processing_expressions`.  These would be define at the item level in the Model Input Object instead of the asset level. In this case I think there would be 3 `processing:expression` objects for each spectral index. We could validate that `raster:bands` and the array of `processing:expression` objects sum to `mlm:input.bands`

**Reply by fmigneault** [reply]
> convert the `pre_processing_function` field into an array of `processing:expression` objects?

One concern I have with this, is that it gets very quickly hard to "align" the indices between 2 sets of arrays. There is no easy way to ensure they match in size. There is also an ambiguity if a single element. Does that mean it is applied to all bands, or one has to explicitly repeat the processing reference for each band? In the example above, that forces us to add `null` to the first 9 bands. And then, if we still want to apply a global pre-processing (eg: custom normalization technique) to all bands after generating them, we can't anymore.


> This field could be called `mlm:processing_expressions`. These would be define at the item level in the Model Input Object instead of the asset level.

An interesting pattern seen in (at least) `storage` and `authentication` extensions is a `<ext>:scheme` at the Item level and a `<ext>:refs` in the places where needed, using a named element from defined schemes. A similar strategy seems to be relevant here, but since we are already listing the `bands`, we could reuse it rather than inserting yet another property. 

I could see the following being a good tradeoff:

```json
{
  "properties": {
    "mlm:processing_expressions": {
      "ndvi": {
        "format": "rio-calc",
        "expression": "(B08 - B04) / (B08 + B04)"
      },
      "ndwi1": {
        "format": "rio-calc",
        "expression": "(B03 - B08) / (B03 + B08)"
      } 
      "ndwi2": {
        "format": "rio-calc",
        "expression": "(B08 - B11) / (B08 + B11)"
      } 
    },
	"mlm:input": [
      {
        "name": "EO Data",
        "bands": [
          {"name": "B01"}, 
		  {"name": "B02"}, 
		  {"name": "B03"}, 
		  {"name": "B04"}, 
		  {"name": "B08"}, 
		  {"name": "B8A"}, 
		  {"name": "B09"}, 
		  {"name": "B11"}, 
		  {"name": "B12"},
		  {"name": "NDVI", "expression": "ndvi"},
		  {"name": "NDWI1", "expression": "ndwi1"},
		  {"name": "NDWI2", "expression": "ndwi2"}
        ],
		"pre_processing_function": {
          "format": "python",
		  "expression": "https://example.com/my_package.py:custom_reshape_and_norm_function"
        }
        "input": {
          "shape": [-1, 12, 1024, 1024],
          "dim_order": ["batch", "channel", "height", "width"],
          "data_type": "uint16"
        }
      }
    ]
  }
}
```
(note: `bands` could allow an array of `oneOf[ string, {name, expression} ]` for backward compatibility)


> We could validate that raster:bands and the array of processing:expression objects sum to mlm:input.bands

I don't think there is a way to do that directly with JSON schema. Even if there is, considering that bands can already come from 3 different references, that definition would be extremely complicated.

**Reply by m-mohr** [reply]
I don't have the capacity right now to dig into this, sorry.

**Reply by rbavery** [reply]
> And then, if we still want to apply a global pre-processing (eg: custom normalization technique) to all bands after generating them, we can't anymore.

Ah right that makes sense. I like the example you've shown with a separate `mlm:processing_expressions` field.

> I don't think there is a way to do that directly with JSON schema. Even if there is, considering that bands can already come from 3 different references, that definition would be extremely complicated

I think referencing the processing expressions in `mlm:input.bands` addresses that concern I had, it unambiguous what the number of band inputs are if the indexes are referenced in `mlm:input.bands`. 

> 		  {"name": "NDVI", "expression": "ndvi"},
> 		  {"name": "NDWI1", "expression": "ndwi1"},
> 		  {"name": "NDWI2", "expression": "ndwi2"}

I could see the use of `expression` to both reference the band calc equation and the unique id of the index to be potentially confusing. instead what do you think of renaming the key to `processing_expression`?

We could also maybe simplify the structure by removing `mlm:processing_expressions` and including each of those expression objects as additional keys for the corresponding input bands. Not sure if this is easy to support when validating with JSON Schema though? I think this would remove redundancy since the processed bands would only need to be named once (rather than `NDVI` and `ndvi`)

```json
{
  "properties": {
	"mlm:input": [
      {
        "name": "EO Data",
        "bands": [
          {"name": "B01"}, 
		  {"name": "B02"}, 
		  {"name": "B03"}, 
		  {"name": "B04"}, 
		  {"name": "B08"}, 
		  {"name": "B8A"}, 
		  {"name": "B09"}, 
		  {"name": "B11"}, 
		  {"name": "B12"},
		  {"name": "NDVI", "processing:expression":  {
                      "format": "rio-calc",
                      "expression": "(B08 - B04) / (B08 + B04)"
                    }},
		  {"name": "NDWI1", "processing:expression":  {
                      "format": "rio-calc",
                      "expression": "(B03 - B08) / (B03 + B08))"
                    }},
		  {"name": "NDWI2", "processing:expression": {
                      "format": "rio-calc",
                      "expression": "(B08 - B11) / (B08 + B11)"
                    } }
        ],
		"pre_processing_function": {
          "format": "python",
		  "expression": "https://example.com/my_package.py:custom_reshape_and_norm_function"
        }
        "input": {
          "shape": [-1, 12, 1024, 1024],
          "dim_order": ["batch", "channel", "height", "width"],
          "data_type": "uint16"
        }
      }
    ]
  }
}

```

**Reply by fmigneault** [reply]
>  renaming the key to processing_expression?

We could. I was taking inspiration on `processing:expression` that uses a nested `expression` object and the same property name in `pre_processing_function` and `post_processing_function`. However, I see now that it can be confusing that sometimes `expression` is the actual expression, and sometimes a reference to one. Maybe another name like `expression_ref` could be relevant.

> We could also maybe simplify the structure by removing mlm:processing_expressions and including each of those expression objects as additional keys for the corresponding input bands.

I fear that representation might cause false interpretation by users. The `processing:expression` key gives the impression that `processing` extension would validate it, but in fact it won't, even if you include the schema in `stac_extensions` (or not for that matter). 

What could work though, is to have it directly like follows :
```json
{
  "bands": [
     {"name": "B01"}, 
	 {"name": "B02"}, 
	 {"name": "B03"}, 
	 {"name": "B04"}, 
	 {"name": "B08"}, 
	 {"name": "B8A"}, 
	 {"name": "B09"}, 
	 {"name": "B11"}, 
	 {"name": "B12"},
	 {"name": "NDVI", "format": "rio-calc", "expression": "(B08 - B04) / (B08 + B04)"},
	 {"name": "NDWI1", "format": "rio-calc", "expression": "(B03 - B08) / (B03 + B08))"},
	 {"name": "NDWI2", "format": "rio-calc", "expression": "(B08 - B11) / (B08 + B11)"}
  ]
}
```

>  remove redundancy since the processed bands would only need to be named once (rather than NDVI and ndvi)

Using the `mlm:processing_expressions` approach, the `expression: ndvi` is not a redundancy. They look like such because of the names I picked, but it could be `{name: NDVI, expression: my-calc1}` instead. The `expression` cross-reference was required to remove the ambiguity about potential name clashes. 

That being said, I like the above representation better. It is easy to interpret directly.

**Reply by rbavery** [reply]
> What could work though, is to have it directly like follows :

I like that example, looks succinct. Would we still not get validation by including the processing expression in this nested object because it's not at the initial property level?

**Reply by fmigneault** [reply]
In this case, I can include the `$ref` to `processing:expression` to perform the validation (though, not much to validate other than `format: string` since `expression: Any` is allowed). The key difference is that `processing:expression` is not shown, so users are not misled by the extensions available in the STAC Item.

**Reply by fmigneault** [reply]
To be more explicit. We can "align" the field to look like `processing:expression`, but we cannot "hack" the `processing` extension and apply it anywhere we want (i.e.: under `mlm:input[].bands`). We must respect its definition that limits its use under the STAC Item properties. 

Using `$ref` to `processing:expression` is a shortcut to make sure the definitions are "aligned", but technically, we could rewrite them explicitly in MLM to achieve the same behavior. The `processing` extension is not included in `stac_extension` (unless some other properties for lineage would actually use it as intended).

**Reply by fmigneault** [reply, marked as answer]
The solution described in https://github.com/crim-ca/mlm-extension/discussions/7#discussioncomment-9276437 is now implemented in https://github.com/crim-ca/mlm-extension/releases/tag/v1.2.0
